include/nbl/builtin/hlsl/impl/tgmath_impl.hlsl

```diff
@@ -0,0 +1,328 @@
+AUTO_SPECIALIZE_TRIVIAL_CASE_HELPER(log_helper, concepts::Scalar<T>, log, T)
+AUTO_SPECIALIZE_TRIVIAL_CASE_HELPER(exp2_helper, concepts::Scalar<T>, exp2, T)
+AUTO_SPECIALIZE_TRIVIAL_CASE_HELPER(exp_helper, concepts::Scalar<T>, exp, T)
+AUTO_SPECIALIZE_TRIVIAL_CASE_HELPER(floor_helper, concepts::FloatingPointScalar<T>, floor, T)
+#undef AUTO_SPECIALIZE_TRIVIAL_CASE_HELPER
+
+template<typename T>
+requires concepts::FloatingPointScalar<T>
+struct pow_helper<T>
+{
+	using return_t = T;
+	static inline return_t __call(const T x, const T y)
+	{
+		return std::pow(x, y);
+	}
+};
+
+template<typename T>
+requires concepts::FloatingPointScalar<T>
+struct isinf_helper<T>
+{
+	using return_t = bool;
+	static inline return_t __call(const T arg)
+	{
+		// GCC and Clang will always return false with call to std::isinf when fast math is enabled,
+		// this implementation will always return appropriate output regardless is fas math is enabled or not
+		using AsUint = typename unsigned_integer_of_size<sizeof(T)>::type;
+		return tgmath_impl::isinf_uint_impl(reinterpret_cast<const AsUint&>(arg));
+	}
+};
+
+template<typename T>
+requires concepts::FloatingPointScalar<T>
+struct isnan_helper<T>
+{
+	using return_t = bool;
+	static inline return_t __call(const T arg)
+	{
+		// GCC and Clang will always return false with call to std::isnan when fast math is enabled,
+		// this implementation will always return appropriate output regardless is fas math is enabled or not
+		using AsUint = typename unsigned_integer_of_size<sizeof(T)>::type;
+		return tgmath_impl::isnan_uint_impl(reinterpret_cast<const AsUint&>(arg));
+	}
+};
+
+template<typename T, typename U>
+requires concepts::FloatingPoint<T> && (concepts::FloatingPoint<T> || concepts::Boolean<T>)
+struct lerp_helper<T, U>
+{
+	using return_t = T;
+	static inline return_t __call(const T x, const T y, const U a)
+	{
+		return glm::mix(x, y ,a);
+	}
+};
+
+#endif
+
+// C++ and HLSL specializations
+
+template<typename FloatingPoint>
+NBL_PARTIAL_REQ_TOP(concepts::FloatingPointScalar<FloatingPoint>)
+struct erf_helper<FloatingPoint NBL_PARTIAL_REQ_BOT(concepts::FloatingPointScalar<FloatingPoint>) >
+{
+	static FloatingPoint __call(NBL_CONST_REF_ARG(FloatingPoint) _x)
+	{
+#ifdef __HLSL_VERSION
+		const FloatingPoint a1 = 0.254829592;
+		const FloatingPoint a2 = -0.284496736;
+		const FloatingPoint a3 = 1.421413741;
+		const FloatingPoint a4 = -1.453152027;
+		const FloatingPoint a5 = 1.061405429;
+		const FloatingPoint p = 0.3275911;
+
+		FloatingPoint sign = sign(_x);
+		FloatingPoint x = abs(_x);
+
+		FloatingPoint t = 1.0 / (1.0 + p * x);
+		FloatingPoint y = 1.0 - (((((a5 * t + a4) * t) + a3) * t + a2) * t + a1) * t * exp(-x * x);
+
+		return sign * y;
+#else
+		return std::erf(_x);
+#endif
+	}
+};
+template<typename FloatingPoint>
+NBL_PARTIAL_REQ_TOP(concepts::FloatingPointScalar<FloatingPoint>)
+struct erfInv_helper<FloatingPoint NBL_PARTIAL_REQ_BOT(concepts::FloatingPointScalar<FloatingPoint>) >
```

Comment:
general thing about ERF and ERFInv, you have what are called "rational fits" and I'm pretty sure a rational for f16, f32 and f64 needs different amounts of coefficients to get 1 ULP precise answers, therefore I'd make different specializations for `float16_t`, etc.

Reply:
@keptsecret

Reply:
> @keptsecret

is the idea that @keptsecret will find the other lower precision implementations?

Reply:
@keptsecret implemented this so i'm forwarding this message to him